src/Workspaces/Core/Portable/FindSymbols/FindReferences/FindReferencesSearchEngine_MapCreation.cs

```diff
@@ -23,44 +26,48 @@ private async Task<ProjectToDocumentMap> CreateProjectToDocumentMapAsync(Project
             {
-                var finalMap = new ProjectToDocumentMap();
+                using var _ = ArrayBuilder<Task<(ImmutableArray<Document>, ISymbol, IReferenceFinder)>>.GetInstance(out var tasks);
 
-                foreach (var kvp in projectMap)
+                foreach (var (project, projectQueue) in projectMap)
                 {
-                    var project = kvp.Key;
-                    var projectQueue = kvp.Value;
-
-                    var documentMap = new DocumentMap();
-
-                    foreach (var symbolAndFinder in projectQueue)
+                    foreach (var (symbol, finder) in projectQueue)
                     {
-                        _cancellationToken.ThrowIfCancellationRequested();
+                        tasks.Add(Task.Run(() =>
+                            DetermineDocumentsToSearchAsync(project, symbol, finder), _cancellationToken));
+                    }
+                }
 
-                        var (symbol, finder) = symbolAndFinder;
+                var finalMap = new ProjectToDocumentMap();
 
-                        var documents = await finder.DetermineDocumentsToSearchAsync(
-                            symbol, project, _documents, _options, _cancellationToken).ConfigureAwait(false);
-                        foreach (var document in documents.Distinct().WhereNotNull())
+                var results = await Task.WhenAll(tasks).ConfigureAwait(false);
+                foreach (var (documents, symbol, finder) in results)
+                {
+                    foreach (var document in documents)
+                    {
+                        if (!finalMap.TryGetValue(document.Project, out var documentMap))
                         {
-                            if (_documents == null || _documents.Contains(document))
-                            {
-                                documentMap.Add(document, symbolAndFinder);
-                            }
+                            documentMap = new DocumentMap();
+                            finalMap.Add(document.Project, documentMap);
                         }
-                    }
-
-                    Contract.ThrowIfTrue(documentMap.Any(kvp1 => kvp1.Value.Count != kvp1.Value.ToSet().Count));
 
-                    if (documentMap.Count > 0)
-                    {
-                        finalMap.Add(project, documentMap);
+                        documentMap.Add(document, (symbol, finder));
                     }
                 }
 
                 return finalMap;
             }
         }
 
+        private async Task<(ImmutableArray<Document>, ISymbol, IReferenceFinder)> DetermineDocumentsToSearchAsync(
+            Project project, ISymbol symbol, IReferenceFinder finder)
+        {
+            var documents = await finder.DetermineDocumentsToSearchAsync(
+                symbol, project, _documents, _options, _cancellationToken).ConfigureAwait(false);
+            var finalDocs = documents.WhereNotNull().Where(
```

Comment:
previously ```.Distinct()``` was being called on the documents.  Should that still be done, or do duplicates not matter/can't happen?

Similar q above to the removed contract where it was checking that the documentMap in the finalMap had no dupes

Reply:
Yeah.  I'm going to add the .Distinct back in.  It might happen, and that would not be desirable.

Reply:
added the contract back in as well.